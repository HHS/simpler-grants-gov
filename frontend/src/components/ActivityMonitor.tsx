"use client";

import { useUser } from "src/services/auth/useUser";

import { useCallback, useEffect, useRef, useState } from "react";

// when a user is logged in, this will track all clicks and key presses and,
// on each action, if the user's auth token is expiring or expired, will refresh the
// token or log the user out accordingly
export function ActivityMonitor() {
  const { user, refreshIfExpiring, refreshIfExpired, logoutLocalUser } =
    useUser();

  const [listening, setListening] = useState(false);
  const handlerRef = useRef<EventListener>(null);

  const refreshTokenIfExpiringOrLogout = useCallback(() => {
    refreshIfExpired()
      .then((expired) => {
        if (expired) {
          logoutLocalUser();
          return;
        }
        return refreshIfExpiring();
      })
      .catch((e) => {
        console.error("Error refreshing token in response to user activity", e);
      });
  }, [refreshIfExpired, refreshIfExpiring, logoutLocalUser]);

  const addHandlers = useCallback(() => {
    if (!listening) {
      document.addEventListener("click", refreshTokenIfExpiringOrLogout);
      document.addEventListener("keydown", refreshTokenIfExpiringOrLogout);
      setListening(true);
    }
  }, [refreshTokenIfExpiringOrLogout, listening]);

  const removeHandlers = useCallback(() => {
    if (!listening) {
      return;
    }
    document.removeEventListener("click", handlerRef.current as EventListener);
    document.removeEventListener(
      "keydown",
      handlerRef.current as EventListener,
    );
    setListening(false);
    handlerRef.current = null;
  }, [listening]);

  // when logged in status changes, remove or add handlers
  useEffect(() => {
    if (!user || !user.token) {
      removeHandlers();
      return;
    }
    addHandlers();
    return () => {
      removeHandlers();
    };
  }, [user, addHandlers, removeHandlers]);

  // whenever we are not listening for activity, and the handler function has been updated
  // update the ref, so that we have the right reference ready when we want to remove the handlers later.
  // we do not want to pick up any changes made while listening, as those changes will not corerspond to the
  // handler that is actually attached to the document.
  // note that if we try to do this in addHandlers it somehow doesn't work right, given all of the dependencies involved.
  useEffect(() => {
    if (refreshTokenIfExpiringOrLogout && !listening) {
      handlerRef.current = refreshTokenIfExpiringOrLogout;
    }
  }, [refreshTokenIfExpiringOrLogout, listening]);

  return null;
}
