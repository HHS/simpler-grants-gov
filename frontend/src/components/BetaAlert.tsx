import { useTranslations } from "next-intl";
import { ReactNode } from "react";
import { GridContainer, Alert as USWDSAlert } from "@trussworks/react-uswds";

type BetaAlertProps = {
  containerClasses?: string;
  heading?: string;
  alertMessage?: string | ReactNode;
};

const BetaAlert = ({
  containerClasses,
  heading,
  alertMessage,
}: BetaAlertProps) => {
  const t = useTranslations("BetaAlert");
  const defaultAlertMessage = t.rich("alert", {
    LinkToGrants: (content) => <a href="https://www.grants.gov">{content}</a>,
  });

  return (
    <div data-testid="beta-alert" className={containerClasses}>
      <GridContainer>
        <USWDSAlert
          type="warning"
          headingLevel="h2"
          heading={heading || t("alertTitle")}
          noIcon
        >
          {alertMessage || defaultAlertMessage}
        </USWDSAlert>
      </GridContainer>
    </div>
  );
};

export default BetaAlert;
