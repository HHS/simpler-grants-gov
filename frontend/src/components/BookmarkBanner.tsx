import { useTranslations } from "next-intl";
import { GridContainer, Alert as USWDSAlert } from "@trussworks/react-uswds";

type BookmarkBannerProps = {
  containerClasses?: string;
  message?: string;
};

const BookmarkBanner = ({ containerClasses, message }: BookmarkBannerProps) => {
  const t = useTranslations("BookmarkBanner");

  const defaultMessage = (
    <>
      {t("message")}
      <br />
      {t.rich("technicalSupportMessage", {
        mailToGrants: (content) => (
          <a href="mailto:simpler@grants.gov">{content}</a>
        ),
      })}
    </>
  );

  return (
    <div data-testid="bookmark-banner" className={containerClasses}>
      <GridContainer>
        <USWDSAlert type="info" headingLevel="h2" heading={t("title")}>
          {message || defaultMessage}
        </USWDSAlert>
      </GridContainer>
    </div>
  );
};

export default BookmarkBanner;
