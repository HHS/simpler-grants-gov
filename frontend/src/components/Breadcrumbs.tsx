import {
  Breadcrumb,
  BreadcrumbBar,
  BreadcrumbLink,
  GridContainer,
} from "@trussworks/react-uswds";

export type Breadcrumb = {
  title: string;
  path: string;
};

export type BreadcrumbList = Breadcrumb[];

type Props = {
  breadcrumbList: BreadcrumbList;
};

const microdata = {
  ol: {
    itemScope: true,
    itemType: "https://schema.org/BreadcrumbList",
  },
  li: {
    itemScope: true,
    itemProp: "itemListElement",
    itemType: "https://schema.org/ListItem",
  },
  a: {
    itemScope: true,
    itemProp: "item",
    itemType: "https://schema.org/WebPage",
  },
};

const Breadcrumbs = ({ breadcrumbList }: Props) => {
  const breadcrumArray = breadcrumbList.map((breadcrumbInfo, i) => {
    return (
      <Breadcrumb
        key={breadcrumbInfo.title + "-crumb"}
        current={i + 1 === breadcrumbList.length}
        {...microdata.li}
      >
        {i + 1 !== breadcrumbList.length ? (
          <BreadcrumbLink
            href={breadcrumbInfo.path}
            itemID={breadcrumbInfo.path}
            {...microdata.a}
          >
            {}
            <span itemProp="name">{breadcrumbInfo.title}</span>
          </BreadcrumbLink>
        ) : (
          <span itemProp="name">{breadcrumbInfo.title}</span>
        )}
        <meta itemProp="position" content={(i + 1).toString()} />
      </Breadcrumb>
    );
  });

  return (
    <GridContainer data-testid="breadcrumb">
      <BreadcrumbBar listProps={{ ...microdata.ol }}>
        {breadcrumArray}
      </BreadcrumbBar>
    </GridContainer>
  );
};

export default Breadcrumbs;
