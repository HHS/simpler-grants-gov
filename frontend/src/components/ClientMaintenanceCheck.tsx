"use client";

import { useFeatureFlags } from "src/hooks/useFeatureFlags";

import { useEffect } from "react";

/**
 * Checks for the "major" features being offline, otherwise we're not in maintenance, so redir to homepage.
 */
export default function ClientMaintenanceCheck() {
  const {
    checkFeatureFlag, // An instance of FeatureFlagsManager
  } = useFeatureFlags();

  useEffect(() => {
    if (
      !checkFeatureFlag("searchOff") &&
      !checkFeatureFlag("opportunityOff") &&
      checkFeatureFlag("authOn")
    ) {
      // This piece does not seem reliable, maybe due to re-render or rehyrdrate cylces where we redirect away before the actual client live feature flag is set
      // redirect("/", RedirectType.push);
    }
  }, [checkFeatureFlag]);

  return null;
}
