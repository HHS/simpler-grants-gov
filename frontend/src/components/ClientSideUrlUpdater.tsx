"use client";

import { useSearchParamUpdater } from "src/hooks/useSearchParamUpdater";

import { useRouter } from "next/navigation";
import { useEffect } from "react";

// This should be used as a way to force a url change from a server component
// supports updating:
// - the full url
// - an individual query param
// - the query string
// Note that if you want to zero out the query or param value, you should explicitly pass "" (empty string)
export function ClientSideUrlUpdater({
  param,
  value,
  url,
  query,
}: {
  param?: string;
  value?: string;
  url?: string;
  query?: string;
}) {
  const router = useRouter();
  const { updateQueryParams, searchParams } = useSearchParamUpdater();
  // if query has been passed in, use that, otherwise pass through the current query
  const updatedQuery = query !== undefined ? query : searchParams.get("query");

  useEffect(() => {
    if (url) {
      router.push(url);
    }
  }, [url, router]);
  useEffect(() => {
    if (param && value !== undefined) {
      updateQueryParams(value, param, updatedQuery);
    }
  }, [param, value, updatedQuery, updateQueryParams]);
  return <></>;
}
