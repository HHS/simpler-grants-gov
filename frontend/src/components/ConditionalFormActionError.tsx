import { ErrorMessage } from "@trussworks/react-uswds";

export function ConditionalFormActionError<
  T extends { [key: string]: string[] },
>({ errors, fieldName }: { errors?: T; fieldName: keyof T }) {
  if (!errors) {
    return;
  }
  const error = errors[fieldName];
  if (error) {
    return <ErrorMessage className="maxw-mobile-lg">{error[0]}</ErrorMessage>;
  }
}
