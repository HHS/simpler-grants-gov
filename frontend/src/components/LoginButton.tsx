import { LOGIN_URL } from "src/constants/auth";
import SessionStorage from "src/services/sessionStorage/sessionStorage";

import { PropsWithChildren } from "react";
import { Button } from "@trussworks/react-uswds";

import { USWDSIcon } from "./USWDSIcon";

export function LoginLink({
  children,
  className,
}: { className?: string } & PropsWithChildren) {
  return (
    <a
      href={LOGIN_URL}
      key="login-link"
      className={className}
      onClick={() => {
        const startURL = `${location.pathname}${location.search}`;
        if (startURL !== "") {
          SessionStorage.setItem("login-redirect", startURL);
        }
      }}
    >
      {children}
    </a>
  );
}

export function LoginButton({
  navLoginLinkText,
}: {
  navLoginLinkText: string;
}) {
  return (
    <LoginLink className="padding-0 text-no-underline text-primary-dark display-flex">
      <Button
        type="button"
        className="usa-nav__link font-sans-2xs display-flex text-normal border-0"
        data-testid="sign-in-button"
      >
        <USWDSIcon
          className="usa-icon margin-right-05 margin-left-neg-05"
          name="login"
          key="login-link-icon"
        />
        {navLoginLinkText}
      </Button>
    </LoginLink>
  );
}
