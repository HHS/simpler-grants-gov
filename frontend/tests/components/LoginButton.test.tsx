import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { LoginButton, LoginLink } from "src/components/LoginButton";

const mockSetItem = jest.fn<void, [string, string]>();

jest.mock("src/services/sessionStorage/sessionStorage", () => {
  return {
    __esModule: true,
    default: {
      setItem: (key: string, value: string): void => mockSetItem(key, value),
    },
  };
});

const mockLocation = {
  pathname: "/test-path",
  search: "?param=value",
};

// Save the original location
const originalLocation = global.location;

describe("LoginButton", () => {
  it("matches snapshot", () => {
    const { container } = render(<LoginButton navLoginLinkText="whatever" />);

    expect(container).toMatchSnapshot();
  });
});

describe("LoginLink", () => {
  beforeEach(() => {
    Object.defineProperty(global, "location", {
      configurable: true,
      value: { ...mockLocation },
      writable: true,
    });

    jest.clearAllMocks();
  });

  afterAll(() => {
    Object.defineProperty(global, "location", {
      configurable: true,
      value: originalLocation,
      writable: true,
    });
  });
  it("matches snapshot", () => {
    const { container } = render(
      <LoginLink>
        <span>hi</span>
      </LoginLink>,
    );

    expect(container).toMatchSnapshot();
  });
  it("should store the current URL in session storage when clicking sign in", async () => {
    render(
      <LoginLink>
        <span>hi</span>
      </LoginLink>,
    );

    const signInButton = screen.getByText("hi");
    await userEvent.click(signInButton);

    expect(mockSetItem).toHaveBeenCalledWith(
      "login-redirect",
      "/test-path?param=value",
    );
  });

  it("should not store URL in session storage if pathname and search are empty", async () => {
    Object.defineProperty(global, "location", {
      value: { pathname: "", search: "" },
    });

    render(
      <LoginLink>
        <span>hi</span>
      </LoginLink>,
    );

    const loginButton = screen.getByText("hi");
    await userEvent.click(loginButton);

    expect(mockSetItem).not.toHaveBeenCalled();
  });
});
