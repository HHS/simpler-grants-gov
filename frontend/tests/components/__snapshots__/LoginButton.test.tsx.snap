// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginButton matches snapshot 1`] = `
<div>
  <a
    class="padding-0 text-no-underline text-primary-dark display-flex"
    href="/api/auth/login"
  >
    <button
      class="usa-button usa-nav__link font-sans-2xs display-flex text-normal border-0"
      data-testid="sign-in-button"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="usa-icon usa-icon margin-right-05 margin-left-neg-05"
        role="img"
      >
        <use
          href="/img.jpg#login"
        />
      </svg>
      whatever
    </button>
  </a>
</div>
`;

exports[`LoginLink matches snapshot 1`] = `
<div>
  <a
    href="/api/auth/login"
  >
    <span>
      hi
    </span>
  </a>
</div>
`;
